### ISSUE TITLE
redirectUrl doesn't work with unlisten

### ISSUE THREAD

**Issue by warpech** [PuppetJs/PuppetJs#67]
Continuation from https://github.com/PuppetJs/PuppetJs/issues/42:

`redirectUrl` pattern used in few of our demo apps doesn't work with page morphing disabled.

The pattern relies on server updating the value of `redirectUrl` property in JSON. Client uses a Custom Element called `puppet-redirect` that is bound to changes of that property. When there is a change of value, it morphs the page into the new URL using Puppet.

This has 2 downsides:

1. Doesn't work after Puppet's `unlisten` method was called ("Disable page morphing" in [starcounter-debug-aid](https://github.com/polyjuice/starcounter-debug-aid))
2. Requires Custom Element

First thing to discuss is what is the desired solution. Naively, I propose that desired solution is to be able to force client to redirect from `.json.cs` to a new URL by page morphing (if Puppet morphing is on) or page reloading (if Puppet morphing is off).

**Comment by miyconst** [comment]
>Naively, I propose that desired solution is to be able to force client to redirect from .json.cs to a new URL by page morphing (if Puppet morphing is on) or page reloading (if Puppet morphing is off).

+1 to this.

I thought that we could somehow use native `http redirect` with status code `300`, but it is impossible when changes come from WebSocket.

Maybe `puppet-dom` should does full page reload instead of ignoring history change when unlisted?

**Comment by tomalec** [comment]
> Maybe puppet-dom should does full page reload instead of ignoring history change when unlisted?

I don't think so. When I disable listening, I usually to ignore the changes, not to postpone reaction. Compare with `.removeEventListener` to any kind of element, and event.

**Comment by warpech** [comment]
> Naively, I propose that desired solution is to be able to force client to redirect from .json.cs to a new URL by page morphing (if Puppet morphing is on) or page reloading (if Puppet morphing is off).

So basically this means that [puppet-dom.js line 28](https://github.com/PuppetJs/puppet-redirect/blob/gh-pages/src/puppet-redirect.html#L28) could be changed from:

```js
if(this.hasAttribute("history")) {
```

To:

```js
if(this.hasAttribute("history") && this.puppet.listening) {
```

if only `<puppet-redirect>` would knew which instance of PuppetJs is listening... 

Currently it does not know that. The `puppet-redirect-pushstate` DOM event is broadcasted to all PuppetJs instances, which is problematic because we may end up with 2 Puppets performing URL morphing.

Could we somehow help `<puppet-redirect>` find the correct Puppet instance?

Maybe this would be the correct way to do it:

```html
<puppet-client>
  <template>
    <!-- the root template should be here -->
    <puppet-redirect></puppet-redirect>
  </template>
</puppet-client>
```

That would allow `<puppet-redirect>` to find the Puppet instance by traversing up the DOM tree. In case it doesn't find `<puppet-client>`, it takes `window.puppet` if it's defined.

WDYT?

**Comment by tomalec** [comment]
+1 for the markup as it semantically explains everything. The only problem is that, this would work for `<puppet-client>` only and not for plain PuppetJS. I think it maight be ok, as we could also create JS-only version of puppet-redirect, that will just take puppet instance as optional parameter, otherwise listen to document events.

Plus for `<puppet-redirect>` we can provide `ref` attribute in the same manner we did for `<puppet-client>` - to support non-nested markup.

**Comment by warpech** [comment]
For plain PuppetJs I propose to use `window.puppet`, overridable by setting `<puppet-redirect>.puppet`
